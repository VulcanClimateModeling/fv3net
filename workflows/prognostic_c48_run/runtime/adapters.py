from typing import Mapping, Set, Sequence, Hashable

import xarray as xr

from fv3fit._shared import Predictor

NameDict = Mapping[Hashable, Hashable]


def _invert_dict(d: Mapping) -> Mapping:
    return dict(zip(d.values(), d.keys()))


class RenamingAdapter:
    """Adapter object for renaming model variables

    Attributes:
        model: a model to rename
        rename_in: mapping from standard names to input names of model
        rename_out: mapping from standard names to the output names of model
    
    """

    def __init__(
        self, model: Predictor, rename_in: NameDict, rename_out: NameDict = None
    ):
        self.model = model
        self.rename_in = rename_in
        self.rename_out = {} if rename_out is None else rename_out

    def _rename(self, ds: xr.Dataset, rename: NameDict) -> xr.Dataset:

        all_names = set(ds.dims) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        redimed = ds.rename_dims(rename_restricted)

        all_names = set(ds.data_vars) & set(rename)
        rename_restricted = {key: rename[key] for key in all_names}
        return redimed.rename(rename_restricted)

    def _rename_inputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, self.rename_in)

    def _rename_outputs(self, ds: xr.Dataset) -> xr.Dataset:
        return self._rename(ds, _invert_dict(self.rename_out))

    @property
    def input_variables(self) -> Set[str]:
        invert_rename_in = _invert_dict(self.rename_in)
        return {invert_rename_in.get(var, var) for var in self.model.input_variables}

    def predict(self, arg: xr.Dataset) -> xr.Dataset:
        input_ = self._rename_inputs(arg)
        prediction = self.model.predict(input_)
        return self._rename_outputs(prediction)


class StackingAdapter:
    """Wrap a model to work with unstacked inputs
    
    Attributes
        model: a model to stack
        sample_dims: sample dimension name of the resulting stacked array
    
    """

    def __init__(self, model: Predictor, sample_dims: Sequence[str]):
        self.model = model
        self.sample_dims = sample_dims

    @property
    def input_variables(self) -> Set[str]:
        return set(self.model.input_variables)

    def predict(self, ds: xr.Dataset) -> xr.Dataset:
        self.model.predict_columnwise(ds, sample_dims=self.sample_dims)
