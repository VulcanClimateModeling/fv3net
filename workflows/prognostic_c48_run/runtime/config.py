from typing import List, Optional
import dataclasses
import yaml
import f90nml

import dacite

from runtime.diagnostics.manager import (
    DiagnosticFileConfig,
    FortranFileConfig,
    get_chunks,
)
from runtime.steppers.nudging import NudgingConfig
from runtime.steppers.machine_learning import MachineLearningConfig

FV3CONFIG_FILENAME = "fv3config.yml"


@dataclasses.dataclass
class UserConfig:
    """The top-level object for python runtime configurations

    Attributes:
        diagnostics: list of diagnostic file configurations
        fortran_diagnostics: list of Fortran diagnostic outputs. Currently only used by
            post-processing and so only name and chunks items need to be specified.
        scikit_learn: a machine learning configuration
        nudging: nudge2fine configuration. Cannot be used if any scikit_learn model
            urls are specified.
        step_tendency_variables: variables to compute the tendencies of.
            These could in principle be inferred from the requested diagnostic
            names.
        step_storage_variables: variables to compute the storage of. Needed for certain
            diagnostics.
    """

    diagnostics: List[DiagnosticFileConfig]
    fortran_diagnostics: List[FortranFileConfig]
    scikit_learn: MachineLearningConfig = MachineLearningConfig()
    nudging: Optional[NudgingConfig] = None
    step_tendency_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(
            ("specific_humidity", "air_temperature", "eastward_wind", "northward_wind",)
        )
    )
    step_storage_variables: List[str] = dataclasses.field(
        default_factory=lambda: list(("specific_humidity", "total_water"))
    )


def get_config() -> UserConfig:
    """Open the configurations for this run
    
    .. warning::
        Only valid at runtime
    """
    with open("fv3config.yml") as f:
        config = yaml.safe_load(f)
    return dacite.from_dict(UserConfig, config)


def get_namelist() -> f90nml.Namelist:
    """Open the fv3 namelist
    
    .. warning::
        Only valid at runtime
    """
    return f90nml.read("input.nml")


def write_chunks(config: UserConfig):
    """Given UserConfig, write chunks to 'chunks.yaml'"""
    diagnostic_file_configs = (
        config.fortran_diagnostics + config.diagnostics  # type: ignore
    )
    chunks = get_chunks(diagnostic_file_configs)
    with open("chunks.yaml", "w") as f:
        yaml.safe_dump(chunks, f)
