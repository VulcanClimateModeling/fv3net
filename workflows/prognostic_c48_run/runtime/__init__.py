from .config import get_namelist, get_config
from .capture import capture_stream, capture_stream_mpi, capture_fv3gfs_funcs
from .diagnostics import (
    get_diagnostic_files,
    DiagnosticFile,
    compute_ml_diagnostics,
    compute_ml_momentum_diagnostics,
    compute_nudging_diagnostics,
    rename_diagnostics,
    default_diagnostics,
)
from .adapters import RenamingAdapter, MultiModelAdapter
from .derived_state import DerivedFV3State
from .nudging import (
    nudging_timescales_from_dict,
    setup_get_reference_state,
    get_nudging_tendency,
    set_state_sst_to_reference,
)

__version__ = "0.1.0"
