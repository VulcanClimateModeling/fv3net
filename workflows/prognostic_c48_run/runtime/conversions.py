import xarray as xr
import pace.util

from runtime.types import QuantityState


def _insert_units_if_necessary(da: xr.DataArray) -> xr.DataArray:
    if "units" not in da.attrs:
        return da.assign_attrs(units="")
    else:
        return da


def dataset_to_quantity_state(ds: xr.Dataset) -> QuantityState:
    quantity_state: QuantityState = {
        variable: pace.util.Quantity.from_data_array(
            _insert_units_if_necessary(ds[variable])
        )
        for variable in ds.data_vars
    }
    return quantity_state


def quantity_state_to_dataset(quantity_state: QuantityState) -> xr.Dataset:
    ds = xr.Dataset(
        {
            variable: quantity_state[variable].data_array
            for variable in quantity_state.keys()
        }
    )
    return ds
