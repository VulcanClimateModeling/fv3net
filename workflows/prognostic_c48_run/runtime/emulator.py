import dataclasses
import xarray as xr
from typing import (
    Callable,
    Hashable,
    Mapping,
    MutableMapping,
    Optional,
    Iterable,
    Set,
    Union,
)

import fv3fit.emulation.thermobasis.emulator
from fv3fit.emulation.thermobasis.xarray import get_xarray_emulator
from runtime.monitor import Monitor
from runtime.names import SPHUM, DELP
from runtime.types import State, Diagnostics, Step

__all__ = ["PrognosticAdapter", "Config"]


def strip_prefix(prefix: str, variables: Iterable[str]) -> Set[str]:
    return {k[len(prefix) :] for k in variables if k.startswith(prefix)}


@dataclasses.dataclass
class Config:
    """
    Attributes:
        emulator: Either a path to a model to-be-loaded or an emulator
            configuration specifying model parameters.
        online: if True, the emulator will be replace fv3 physics for all
            humidities above level ``ignore_humidity_below``.
        train: if True, each timestep will be used to train the model.
        ignore_humidity_below: see ``online``.
    """

    emulator: Union[
        str, fv3fit.emulation.thermobasis.emulator.Config
    ] = dataclasses.field(default_factory=fv3fit.emulation.thermobasis.emulator.Config)
    # online parameters
    online: bool = False
    train: bool = True
    # will ignore the emulator for any z larger than this value
    # remember higher z is lower in the atmosphere hence "below"
    ignore_humidity_below: Optional[int] = None


@dataclasses.dataclass
class PrognosticAdapter:
    """Wrap a Step function with an emulator

    The wrapped function produces diagnostic outputs prefixed with
    ``self.emulator_prefix_`` and trains/applies the emulator to ``state``
    depending on the user configuration.

    Attributes:
        state: The mutable state being updated.
        monitor: A Monitor object to use for saving outputs.
        emulator_prefix: the prefix to use adjust the outputted variable names.
        diagnostic_variables: the user-requested diagnostic variables, will be
            searched for inputs starting with ``emulator_prefix``.
        timestep: the model timestep in seconds
        inputs_to_save: the set of diagnostics that will be produced by the
            emulated step function.
    
    """

    config: Config
    state: State
    emulator_prefix: str = "emulator_"
    diagnostic_variables: Set[str] = dataclasses.field(default_factory=set)
    timestep: float = 900

    def __post_init__(self: "PrognosticAdapter"):
        self.emulator = get_xarray_emulator(self.config.emulator)

    @property
    def monitor(self) -> Monitor:
        return Monitor.from_variables(
            self.diagnostic_variables, self.state, self.timestep
        )

    @property
    def inputs_to_save(self) -> Set[str]:
        return set(strip_prefix(self.emulator_prefix, self.diagnostic_variables))

    def emulate(self, name: str, func: Step) -> Diagnostics:
        inputs = {key: self.state[key] for key in self.emulator.input_variables}

        inputs_to_save: Diagnostics = {
            self.emulator_prefix + key: self.state[key] for key in self.inputs_to_save
        }
        before = self.monitor.checkpoint()
        diags = func()
        change_in_func = self.monitor.compute_change(name, before, self.state)

        if self.config.train:
            self.emulator.partial_fit(inputs, self.state)
        emulator_prediction = self.emulator.predict(inputs)

        emulator_after: MutableMapping[Hashable, xr.DataArray] = {
            DELP: before[DELP],
            **emulator_prediction,
        }

        # insert state variables not predicted by the emulator
        for v in before:
            if v not in emulator_after:
                emulator_after[v] = self.state[v]

        changes = self.monitor.compute_change("emulator", before, emulator_after)

        if self.config.online:
            mask = compute_mask(self.config.ignore_humidity_below)
            update_state_with_emulator(self.state, emulator_prediction, mask)
        return {**diags, **inputs_to_save, **changes, **change_in_func}

    def __call__(self, name: str, func: Step) -> Step:
        """Emulate a function that updates the ``state``
        
        Similar to :py:class:`runtime.monitor.Monitor` but with prognostic emulation
        capability.

        Args:
            name: The name of the emulator
            func: a function that updates the State and returns a dictionary of
                diagnostics (usually a method of :py:class:`runtime.loop.TimeLoop`).
        
        Returns:
            emulated_func: a function which observes the change to
                ``self.state`` done by ``func`` and optionally applies/trains an ML
                emulator.

        """

        def step() -> Diagnostics:
            return self.emulate(name, func)

        # functools.wraps modifies the type and breaks mypy type checking
        step.__name__ = func.__name__

        return step


def update_state_with_emulator(
    state: MutableMapping[Hashable, xr.DataArray],
    src: Mapping[Hashable, xr.DataArray],
    compute_mask: Callable[[Hashable, xr.DataArray], xr.DataArray],
) -> None:
    """
    Args:
        state: the mutable state object
        src: updates to put into state
        compute_mask: a function returning a mask. Where this mask is True, the
            original state array will be used.

    """
    for key in src:
        arr = state[key]
        mask = compute_mask(key, arr)
        state[key] = arr.where(mask, src[key].variable)


@dataclasses.dataclass
class compute_mask:
    ignore_humidity_below: Optional[int] = None

    def __call__(self, name: Hashable, arr: xr.DataArray) -> xr.DataArray:
        if name == SPHUM:
            if self.ignore_humidity_below is not None:
                return arr.z > self.ignore_humidity_below
            else:
                return xr.DataArray(False)
        else:
            return xr.DataArray(True)
