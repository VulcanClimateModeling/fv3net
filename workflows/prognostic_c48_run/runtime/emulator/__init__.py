from fv3fit.emulation.thermobasis.emulator import (
    Config as OnlineEmulatorConfig,
    Emulator,
    BatchDataConfig,
)
from fv3fit.emulation.thermobasis.loss import QVLoss, MultiVariableLoss, RHLoss
