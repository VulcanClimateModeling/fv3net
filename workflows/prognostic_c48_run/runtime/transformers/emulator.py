import dataclasses
from typing import Optional, Union, Sequence

import fv3fit.emulation.thermobasis.emulator
from fv3fit.emulation.thermobasis.xarray import get_xarray_emulator
from runtime.masking import get_mask, where_masked
from runtime.types import State

__all__ = ["Config", "Adapter"]


@dataclasses.dataclass
class Config:
    """
    Attributes:
        emulator: Either a path to a model to-be-loaded or an emulator
            configuration specifying model parameters.
        online: if True, the emulator will be replace fv3 physics for all
            humidities above level ``ignore_humidity_below``.
        train: if True, each timestep will be used to train the model.
        mask_kind: the type of mask. Defers to the functions
            ``runtime.masking.compute_{mask_kind}``. The default does not mask any
            of the emulator predictions.
        ignore_humidity_below: if mask_kind ="default", then use the fv3 physics
            instead of the emulator above this level.
    """

    emulator: Union[
        str, fv3fit.emulation.thermobasis.emulator.Config
    ] = dataclasses.field(default_factory=fv3fit.emulation.thermobasis.emulator.Config)
    # online parameters
    online: bool = False
    train: bool = True
    # will ignore the emulator for any z larger than this value
    # remember higher z is lower in the atmosphere hence "below"
    ignore_humidity_below: Optional[int] = None
    mask_kind: str = "default"


@dataclasses.dataclass
class Adapter:
    config: Config

    def __post_init__(self: "Adapter"):
        self.emulator = get_xarray_emulator(self.config.emulator)

    def predict(self, inputs: State) -> State:
        return self.emulator.predict(inputs)

    def apply(self, prediction: State, state: State):
        if self.config.online:
            updated_state = where_masked(
                state,
                prediction,
                compute_mask=get_mask(
                    self.config.mask_kind, self.config.ignore_humidity_below
                ),
            )
            state.update(updated_state)

    def partial_fit(self, inputs: State, state: State):
        if self.config.train:
            self.emulator.partial_fit(inputs, state)

    @property
    def input_variables(self) -> Sequence[str]:
        return self.emulator.input_variables
