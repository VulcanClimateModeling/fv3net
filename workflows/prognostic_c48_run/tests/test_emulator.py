import tensorflow as tf
import joblib
from runtime.emulator import (
    PrognosticAdapter,
    Config,
)
from fv3fit.emulation.thermobasis.emulator import Config as MLConfig


def regression_state(state, regtest):
    for v in sorted(state):
        print(v, joblib.hash(state[v].values), file=regtest)


def test_state_regression(state, regtest):
    regression_state(state, regtest)


def test_adapter_regression(state, regtest):
    tf.random.set_seed(0)

    name = "add_one"

    emulate = PrognosticAdapter(
        Config(MLConfig(levels=state["air_temperature"].z.size)),
        state,
        diagnostic_variables={
            "emulator_latent_heat_flux",
            "tendency_of_specific_humidity_due_to_emulator",
            f"tendency_of_specific_humidity_due_to_{name}",
        },
        timestep=900,
    )

    def add_one_to_temperature():
        state["air_temperature"] += 1
        return {"some_diag": state["specific_humidity"]}

    out = emulate(name, add_one_to_temperature)()

    # sort to make the check deterministic
    regression_state(out, regtest)
