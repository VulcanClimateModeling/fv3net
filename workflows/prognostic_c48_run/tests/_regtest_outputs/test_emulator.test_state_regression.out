accumulated_x_courant_number d8e240d6051b6989d649c9d801fa5ce0
accumulated_x_mass_flux 78d7bbd05e535379f567d4c1a4f1dd68
accumulated_y_courant_number 46487434ec6e4a9161ada0a2c4258090
accumulated_y_mass_flux 6e57095c43b1cd49bf4e683402011731
air_temperature 2ef20f383fc58d3c0280a70ae0bfed49
air_temperature_after_physics d32e9479916f5343b8d8e99163b6cd81
air_temperature_at_2m 93189c5f15db009b5d1852b1446e90a8
area_of_grid_cell 730cafe4b44079e1752b3599d3ef7575
atmosphere_hybrid_a_coordinate 448c691d6f2f72d52d71347cfa1c0ebb
atmosphere_hybrid_b_coordinate 59163fb88e42052b0c6fbd5e2691a65e
canopy_water 0ca575c71f3e9125e63ddcf4cdf78f78
clear_sky_downward_longwave_flux_at_surface c748953ce040190fb68ac2beff0a57b1
clear_sky_downward_shortwave_flux_at_surface 036acbcbe423468c3a61b44b417cb6da
clear_sky_upward_longwave_flux_at_surface 0b094e821b97df30b822965c3745155c
clear_sky_upward_longwave_flux_at_top_of_atmosphere 9070d1287eb49b704fa0cb3e35b2a1eb
clear_sky_upward_shortwave_flux_at_surface 036acbcbe423468c3a61b44b417cb6da
clear_sky_upward_shortwave_flux_at_top_of_atmosphere 036acbcbe423468c3a61b44b417cb6da
cloud_amount 335d6dcd7eb30a99b1a98a1a3468fca2
cloud_ice_mixing_ratio d4793ced39fed4d2c47d1870536734a4
cloud_water_mixing_ratio c61b90a8e859620a686d72ce908c21aa
convective_cloud_bottom_pressure 036acbcbe423468c3a61b44b417cb6da
convective_cloud_fraction 036acbcbe423468c3a61b44b417cb6da
convective_cloud_top_pressure 036acbcbe423468c3a61b44b417cb6da
cos_zenith_angle f458777057a1dac1dcabaeccbb09d481
deep_soil_temperature c47d11e4610f7c2985bf523c1ad3fc47
dissipation_estimate_from_heat_source e305dd8f07f4fc9e68d4c2efe5994f63
eastward_wind 37f60490010060d52753aaefd5bc2970
eastward_wind_after_physics 355e65c61e37cdd0d8e1fe787d12efe0
eastward_wind_at_surface 3c9ad9f79812accdd074cc581e28d823
evaporation 435a6571474f7b3076795683dbb61c08
fh_parameter c2416ca40d5773e59dd5df8f2b6d02a7
fm_at_10m 5b6cf7796c0e64d4758ce62ac59e81f6
fm_parameter a53537350516bb32fd5fe353b9a10b12
fractional_coverage_with_strong_cosz_dependency 289c5daff5a739e181211943e50fb693
fractional_coverage_with_weak_cosz_dependency 4635cb2b03c5fa080ed14d4d4488db87
friction_velocity f97bdb2a4210fa1fbcf16f7ad096fa92
graupel_mixing_ratio abffb9ebf88072f8782d0bb6fb09864f
ice_fraction_over_open_water 036acbcbe423468c3a61b44b417cb6da
interface_pressure 94a6f6197a7aed920d3dae2aa918df42
interface_pressure_raised_to_power_of_kappa 10964c2db29185e5ced668b441f61f09
land_sea_mask 0803c7b71702e3d204a2adff11ad2bd3
land_surface_temperature 031689915cf676e8df3051e9ff845788
latent_heat_flux ad95faf8073cd26d0be4cf7d0abd7c5f
latitude a32966cfa6c4619564e8a80572070d4b
layer_mean_pressure_raised_to_power_of_kappa 8a2a232a0bf43af4f39ef4e9c4410991
liquid_soil_moisture 572c88452d7258f7f02ab6a0a0118a3d
logarithm_of_interface_pressure aa820210a25b76ecf0e9d37bb405d478
longitude 230b12c7e20bf84803437aa0913fdd61
maximum_fractional_coverage_of_green_vegetation 66f415692412ac86058a836186088344
maximum_snow_albedo_in_fraction 022812bc12dbd3f8153fc8532eb861de
mean_cos_zenith_angle 036acbcbe423468c3a61b44b417cb6da
mean_near_infrared_albedo_with_strong_cosz_dependency 75a207a27e0c8f5326101a3f0cb69a6b
mean_near_infrared_albedo_with_weak_cosz_dependency 657ea13549685727dbe52fedf301651f
mean_visible_albedo_with_strong_cosz_dependency 8463f9874a862e5c71184e4023082439
mean_visible_albedo_with_weak_cosz_dependency 20768ffce38602057965d82fe2836b8b
minimum_fractional_coverage_of_green_vegetation 454d53197d335d8f670033f949c1db6d
northward_wind 127e4f2e7a2a01354410a6db14cb8551
northward_wind_after_physics de4fffd9b7db49d4b00eaea421eec1df
northward_wind_at_surface 114db718a709571bd0843f8897e455e2
ocean_surface_temperature 3b84a6c474cb4c3cf6a61ab9c29bc85c
ozone_mixing_ratio 137559c73b79aec26a33cbe08c443ea8
pressure_thickness_of_atmospheric_layer da12de70e35aa93947c66e4fe3bcd258
rain_mixing_ratio 48a8b9cceb4950523444076eebf65922
sea_ice_thickness 036acbcbe423468c3a61b44b417cb6da
sensible_heat_flux 0ae5a4b34f91e123a05c1a662cb82b5c
snow_cover_in_fraction 036acbcbe423468c3a61b44b417cb6da
snow_depth_water_equivalent 036acbcbe423468c3a61b44b417cb6da
snow_mixing_ratio 2962038863f3da0c97d88072a24c498c
snow_rain_flag 036acbcbe423468c3a61b44b417cb6da
soil_temperature 36563492ee07f21059903fba16964539
soil_type 9c864af45c4c15fad6e393f9e57f2747
specific_humidity 7adcb91ffc7469fc129c6b3a247cfcb1
specific_humidity_at_2m 6ca985eb929c8d3f24b585f0f33177fd
surface_geopotential d426df12afc4b0f913e51bb44f0d87c9
surface_pressure 93dc7845c1c169c3511d02c0f1a8ddc7
surface_roughness 509f557520b7f59468b1f272961b363d
surface_slope_type c2e94f105de64c126f32abe3ddefdee4
surface_temperature 031689915cf676e8df3051e9ff845788
surface_temperature_over_ice_fraction 031689915cf676e8df3051e9ff845788
total_condensate_mixing_ratio 1735f910655840737e57ccc0f8cb60b2
total_precipitation 036acbcbe423468c3a61b44b417cb6da
total_sky_downward_longwave_flux_at_surface b1d389164bf0de6b0c6382360f3b396a
total_sky_downward_shortwave_flux_at_surface 036acbcbe423468c3a61b44b417cb6da
total_sky_downward_shortwave_flux_at_top_of_atmosphere 036acbcbe423468c3a61b44b417cb6da
total_sky_upward_longwave_flux_at_surface 996bddbfbe54c415b898651581456355
total_sky_upward_longwave_flux_at_top_of_atmosphere 30440572181fc5515a23ab933814c9b0
total_sky_upward_shortwave_flux_at_surface 036acbcbe423468c3a61b44b417cb6da
total_sky_upward_shortwave_flux_at_top_of_atmosphere 036acbcbe423468c3a61b44b417cb6da
total_soil_moisture 572c88452d7258f7f02ab6a0a0118a3d
total_water 9b049528fc89a027036aca8d99554043
vegetation_fraction 49cfa474150b0f059ec7ccdc834c0894
vegetation_type ff691aa48c5a6a3cd5e9f15341d28102
vertical_pressure_velocity 498f1a5d2ff96100c71f81de46045dc3
vertical_thickness_of_atmospheric_layer 7ee6c48a9affd1ad17c98d965ad7e176
vertical_wind 703a83e07aeef63d8ef8d89df6ddaa44
water_equivalent_of_accumulated_snow_depth 036acbcbe423468c3a61b44b417cb6da
x_wind d679c87246478bd14c525595729c8a5f
x_wind_on_c_grid 948ebe2bf682da5da4b8a1d79da4c6f4
y_wind 049dd6ad594bb20efc65658e869d14ac
y_wind_on_c_grid 6a485abce16d191d2444f96ae6113a98
